TYP: `_lib.doccer`: add type annotations

pavyamsiri opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes gh-20854.

#### What does this implement/fix?
<!--Please explain your changes.-->
This adds type annotations to _lib.doccer along with some updates to the function docstrings.

#### Additional information
I added a Decorator type alias to avoid having to repeat Callable[Callable[...], ...] etc.

pavyamsiri committed:
doccer: Change single quote strings to double quote

pavyamsiri committed:
doccer: Add type annotations to docformat

Change string to str to reflect the actual python type

pavyamsiri committed:
doccer: Type annotate indentcount_lines

Update docstring as well.

pavyamsiri committed:
doccer: Add TypeAlias for Decorator

pavyamsiri committed:
doccer: Add type annotation to inherit_docstring_from

Additionally update docstring

pavyamsiri committed:
doccer: Add type annotations for extent_notes_in_docstring

Update the docstring

pavyamsiri committed:
doccer: Add type annotations for replace_notes_in_docstring

Update docstrings

pavyamsiri committed:
doccer: Add type annotations for unindent_string

Update docstring

pavyamsiri committed:
doccer: Add type annotations to unindent_dict

Update docstring

pavyamsiri committed:
doccer: Add type annotation to filldoc

Update docstring

pavyamsiri committed:
doccer: Handle None case for filldoc

__doc__ can be None so we have to set a default

pavyamsiri committed:
doccer: Add type annotation to doc_replace

Update docstring

pavyamsiri committed:
doccer: Add type annotation to dict in unindent_dict

Pyright complains about this.

pavyamsiri committed:
doccer: Fix up docstrings

pavyamsiri committed:
DOC: In doccer avoid naming return values

lucascolley commented:
The CI failures are relevant - you can use `python dev.py lint` and `python dev.py mypy` to see them locally.

pavyamsiri committed:
DOC: Wrap the docstring to not go past character limit

pavyamsiri committed:
TYP: Add type annotations to docfiller variables

pavyamsiri committed:
TYP: Change Decorator type alias to allow for methods

The old type alias did not allow methods to be decorated.

pavyamsiri committed:
TYP: Add back in ParamSpec

It is still used in the inner decorator functions.

pavyamsiri commented:
Thanks for the tip on running it locally. I think I have fixed all the issues now.

rgommers commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
-        The string to replace from the original docstring.
-    newval : string
-        The string to replace ``oldval`` with.
+    obj : type[Any] | Any
```
This kind of change is typically undesirable. Docstrings are first of all for humans to read; the goal is to keep this correct and readable, not to duplicate the annotations from the signature. The old `object` was file here.

pavyamsiri replied:
Thanks for the feedback. I have changed all of these types into just object in the docstrings.

rgommers commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
-        decorator that applies dictionary to input function docstring
-
-    '''
+    Decorator[R]
```
`decfunc : function` is correct; the numpydoc format is `name : type`. Can you undo changes like these?

rgommers reviewed (commented):
Thanks @pavyamsiri. The added type annotations look quite good, and most of the added docs look good too. The one issue that jumped out at me is the replacing of parameters/returns descriptions with type annotations, which isn't right. I commented on two instances of those, I hope that clears up the intent. That applies more generally to docstring changes in this PR.

pavyamsiri committed:
DOC: Fix docstrings in _lib.doccer

Avoid duplicating type annotations in docstrings and use more readable
types.

pavyamsiri commented:
@rgommers Sorry I know you are probably busy but would it possible to re-review this PR? I think everything should be okay now. Although I'm not sure if there were subsequent PRs that made this obsolete.

Just wanted to make sure if I should keep this PR open.

tylerjereddy commented:
In typing land, we could also ask @jorenham  :)

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -54,24 +67,25 @@ def docformat(docstring, docdict=None):
-    """
-    This decorator modifies the decorated function's docstring by
+def inherit_docstring_from(
+    cls: type[Any] | Any,
```
The [typing docs](https://typing.readthedocs.io/en/latest/spec/concepts.html#union-types) specify `T | Any` to mean something very specific:

> The union `T | Any` is not reducible to a simpler form. It represents an unknown static type with lower bound `T`. That is, it represents an unknown set of objects which may be as large as `object`, or as small as `T`, but no smaller.

But I don't think that mypy or pyright currently implement it this way, and consider `T | Any` equivalent to `Any`. 
However, this might change in the near future. So that's why I tend to avoid using `T | Any`. 

---

As a sidenote; there are more ways in which `Any` can lead to unexpected outcomes, so in my own projects I usually prefer `object` if it's possible. In this case, it shouldn't matter too much; this mostly concerns return types and type arguments.

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -1,26 +1,39 @@
+Decorator: TypeAlias = Callable[[Callable[..., R]], Callable[..., R]]
+
+
+def docformat(docstring: str, docdict: dict[str, str] | None = None) -> str:
```
In case you also want to accept things like `TypedDict` instances, or immutable dict-likes, you could use `Mapping[str, str]` from `collections.abc` instead.

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -1,26 +1,39 @@
 
-def docformat(docstring, docdict=None):
-    ''' Fill a function docstring from variables in dictionary
+Decorator: TypeAlias = Callable[[Callable[..., R]], Callable[..., R]]
```
The parameter signatures of the passed functions will get lost here. 
If this decorator type doesn't change the function signature (by e.g. changing the type of some parameter), then you instead could use something like 

```suggestion
_F = TypeVar("_F", bound=Callable[..., object])

Decorator: TypeAlias = Callable[[_F], _F]
```

The reason for binding the entire callable, and not using `ParamSpec`, is that this way the entire function type is bound, including all overloads, and possible attributes. So this way, e.g. decorating a `np.ufunc` would return that same `np.ufunc` type, instead of effectively upcasting it to a `Callable[..., object]`.

jorenham replied:
Unless I'm missing something, if you apply this "trick" to other similar cases in this module, this would allow you to get rid of the `ParamSpec`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -112,9 +126,9 @@ def inherit_docstring_from(cls):
     """
-    def _doc(func):
+
+    def _doc(func: Callable[P, R]) -> Callable[P, R]:
```
The same reasoning as for `Decorator` applies here; this could be made even better by binding the entire `Callable` as `def _doc(func: _F) -> _F`. 

And if wou really want to dot the "i"'s here, you could consider making `func` positional-only

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -123,63 +137,112 @@ def _doc(func):
     """
-    def _doc(func):
+
+    def _doc(func: Callable[P, R]) -> Callable[P, R]:
```
See my earlier comment about transforming this to `func: _F`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -1,26 +1,39 @@
+    "doc_replace",
 ]
 
+P = ParamSpec("P")
+R = TypeVar("R")
```
Unless you use these typevar-likes in other modules too (which I wouldn't recommend), then the convention is to prefix their names with a `"_'"`, i.e.

```suggestion
_P = ParamSpec("_P")
_R = TypeVar("_R")
```

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -80,14 +94,14 @@ def inherit_docstring_from(cls):
     Parameters
     ----------
-    cls : Python class or instance
+    cls : object
```
Technically a `type` is also and `object`, but it maybe it could help to explicitly include `type` anyway

pavyamsiri replied:
I'm not exactly sure what to put here then, something like `object | type` or just `type`?

jorenham replied:
yea something like that

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -123,63 +137,112 @@ def _doc(func):
-def extend_notes_in_docstring(cls, notes):
-    """
-    This decorator replaces the decorated function's docstring
+def extend_notes_in_docstring(cls: type[Any] | Any, notes: str) -> Decorator[R]:
```
See my earlier comments about `type[Any]` and `T | Any`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -123,63 +137,112 @@ def _doc(func):
-def replace_notes_in_docstring(cls, notes):
-    """
-    This decorator replaces the decorated function's docstring
+def replace_notes_in_docstring(cls: type[Any] | Any, notes: str) -> Decorator[R]:
```
See my earlier comments about `type[Any]` and `T | Any`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -123,63 +137,112 @@ def _doc(func):
 
-def indentcount_lines(lines):
-    ''' Minimum indent for all lines in line list
+def indentcount_lines(lines: Iterable[str]) -> int:
```
I can't think of a better way to annotate this, and maybe you alreayd know this; but since `str` itself is also an `Iterable[str]`, this also accepts `lines: str` as argument 🤷🏻

pavyamsiri replied:
I actually didn't know that or think of that but yes that makes sense that str is Iterable[str] as well. If so I'm not sure of a way to type annotate that we want something that iterates over str but it can't be str without doing a runtime check which we shouldn't do.

jorenham replied:
For sequences it's also the same story; I've seen some people use `list[str] | tuple[str, ...]` in such cases.
But personally I'd rather have parameter type annotations that are too broad instead of too narrow 🤷🏻

jorenham replied:
Anyway, I just wanted to point this out I guess; there's no need to change this as far as I'm concerned.

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
 
-def filldoc(docdict, unindent_params=True):
-    ''' Return docstring decorator using docdict variable dictionary
+def filldoc(docdict: dict[str, str], unindent_params: bool = True) -> Decorator[R]:
```
See my earlier comment about using `Mapping` here

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -123,63 +137,112 @@ def _doc(func):
     """
-    def _doc(func):
+
+    def _doc(func: Callable[P, R]) -> Callable[P, R]:
```
See my earlier comment about transforming this to `func: _F`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
-    def decorate(f):
-        f.__doc__ = docformat(f.__doc__, docdict)
-        return f
+    def decorate(func: Callable[P, R]) -> Callable[P, R]:
```
See my earlier comment about transforming this to `func: _F`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
-def unindent_dict(docdict):
-    ''' Unindent all strings in a docdict '''
-    can_dict = {}
+def unindent_dict(docdict: dict[str, str]) -> dict[str, str]:
```
See my earlier comment about using `Mapping` here for the parameter.
The `dict[str, str]` return type is perfectly fine 👌🏻

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
 
 
-def doc_replace(obj, oldval, newval):
+def doc_replace(obj: type[Any] | Any, oldval: str, newval: str) -> Decorator[R]:
```
See my earlier comments about `type[Any]` and `T | Any`

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
+    doc = (obj.__doc__ or "").replace(oldval, newval)
 
-    def inner(func):
+    def inner(func: Callable[P, R]) -> Callable[P, R]:
```
See my earlier comment about transforming this to `func: _F`

jorenham commented on scipy/io/matlab/_miobase.py:
```diff
@@ -79,7 +79,7 @@ class MatReadWarning(UserWarning):
    If True, write strings as Unicode, else MATLAB usual encoding.'''}
 
-docfiller = doccer.filldoc(doc_dict)
+docfiller: doccer.Decorator = doccer.filldoc(doc_dict)
```
The type argument is missing here. 
But it's not needed to use `doccer.Decorator` here; static type-checkers don't require it. 
So you could just write 

```suggestion
docfiller: Final = doccer.filldoc(doc_dict)
```

(assuming that `typing.Final` is imported)

jorenham commented on scipy/ndimage/_ni_docstrings.py:
```diff
@@ -205,4 +205,4 @@
     }
 
-docfiller = doccer.filldoc(docdict)
+docfiller: doccer.Decorator = doccer.filldoc(docdict)
```
See my comment in `_miobase.py`

jorenham reviewed (commented):
It's good to see type annotations being added to scipy!
At the very least, it could save me some work in writing [`scipy-stubs`](https://github.com/jorenham/scipy-stubs) 😄 .

Don't be alarmed by the amount of comments I left; most of them refer to other ones: It effectively boils down to 3 or 4 suggestions.

pavyamsiri committed:
Merge branch 'scipy:main' into doccer-update

pavyamsiri committed:
TYP: Use Mapping[str, str] instead of dict[str, str]

This only changes in scipy/_lib/doccer.py. This was suggested by
@jorenham.

pavyamsiri committed:
TYP: Change type[Any] | Any to just object.

As suggested by @jorenham as alternative that avoids using Any.

pavyamsiri committed:
TYP: Let Decorator be a Protocol instead of TypeAlias

As suggested by @jorenham, this removes the need for a redundant return
type while also preserving the type of the decorated function.

pavyamsiri committed:
TYP: Use typing.Final instead of doccer.Decorator

When type annotating the type of doccer.filldoc, use typing.Final
instead of doccer.Decorator. This was suggested by @jorenham.

pavyamsiri commented:
@jorenham Thanks for the feedback. I added your suggestions to the PR. Hopefully the CI passes because for some reason on my end, the local lint test seems to find some issues in unrelated code that doesn't happen on my main branch.

Might need to still discuss what to do about `indentcount_lines` and type annotating `lines: Iterable[str]` and also the docstring type for `cls : object`.

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -54,24 +68,25 @@ def docformat(docstring, docdict=None):
-def inherit_docstring_from(cls):
-    """
-    This decorator modifies the decorated function's docstring by
+def inherit_docstring_from(
+    cls: object,
+) -> Decorator:
+    """This decorator modifies the decorated function's docstring by
```
I think that this could fit on one line now

jorenham commented:
> Might need to still discuss what to do about `indentcount_lines` and type annotating `lines: Iterable[str]` and also the docstring type for `cls : object`.

I suppose that the `lines: Iterable[str]` is fine as it is, because there's not really a better alternative. And in the worst case, someone will pass a bare `str`, and that will be accepted by the implementation. So technically it's valid to pass a bare `str`. 

I think that `cls: object` is already an improvement over `cls: type[Any] | Any`. And although `type | object` is technically equivalent to `object`, I think that it could help to annotate it like that either way. Because type annotations are in a way a form of documentation, and "explicit is better than implicit".

jorenham commented on scipy/_lib/doccer.py:
```diff
@@ -1,26 +1,40 @@
-def docformat(docstring, docdict=None):
-    ''' Fill a function docstring from variables in dictionary
+
+class Decorator(Protocol):
```
Will this be part of the public api? If so, then it probably needs a docstring. Otherwise, it should probably be prefixed with a `_`.

lucascolley replied:
This file is under `_lib` so will not be a part of the public API. I think it doesn't really matter either way whether a leading underscore is used here or not ([IMO PEP8 is a bit ambiguous](https://github.com/astral-sh/ruff/issues/10755)). A one-line (or more, if there is more to say) docstring is always nice, though :)

jorenham replied:
Ah of course, we're already operating within the covers of the `_` prefix; my bad 🙃

pavyamsiri replied:
Thanks for reminding me about docstrings! I added a simple one because I'm not entirely sure how to properly document such simple protocols. Tell me if you want me to be more informative and less self-referential (I used the word decorator to describe the decorator protocol haha).

I feel like decorators should just be part of the typing module instead of having to create a custom protocol.

jorenham replied:
Maybe it'll help to point out that this protocol is (only) meant for static typing purposes. Then I suppose it should be clear enough 👌🏻.

> I feel like decorators should just be part of the typing module instead of having to create a custom protocol.

Yea I feel you you... Unfortunately, "DRY" is a virtually impossible to achieve within python typing in its current form

pavyamsiri committed:
STY: Change function def from taking 3 lines to 1 line

pavyamsiri committed:
DOC: Change type annotation for cls to also include `type`

pavyamsiri committed:
DOC: Write a simple docstring for Decorator in _lib/doccer.py.

lucascolley commented on scipy/_lib/doccer.py:
```diff
@@ -80,14 +95,14 @@ def inherit_docstring_from(cls):
     Parameters
     ----------
-    cls : Python class or instance
+    cls : type | object
```
I think we typically write `or` instead of `|` - at least that it was appears in [the style guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters)

jorenham replied:
Ah I wasn't aware. In that case, you should probably use `or` 🤷🏻

pavyamsiri replied:
Should be fixed now.

lucascolley reviewed (commented):
are all of your comments addressed @jorenham ?

jorenham commented:
> are all of your comments addressed @jorenham ?

They are 😄

pavyamsiri committed:
STY: Use or instead of | when writing type annotations.

Thanks to @lucascolley for the reminder.

lucascolley reviewed (approved):
thanks a lot for your first contribution to SciPy @pavyamsiri and for the review @jorenham, let's get this in!

PRs like this are very welcome in the future, the bottleneck is just reviewer time!

jorenham commented:
> PRs like this are very welcome in the future, the bottleneck is just reviewer time!

I wouldn't mind helping review other typing-related PR's.

lucascolley commented:
> I wouldn't mind helping review other typing-related PR's.

Fantastic! We have another first-time contribution waiting for review at gh-20307, if you could take a look at that? Perhaps @pavyamsiri would be interested in taking a look at other modules, too :)
